kernel/src/history_manager/mod.rs

```diff
@@ -1,4 +1,308 @@
+            "LogSegment should ensure that a segment is non-empty"
+        );
+        debug_assert!(log_segment.end_version == snapshot.log_segment().end_version);
+        Ok(Self {
+            log_segment,
+            snapshot,
+        })
+    }
+
+    /// Gets the timestamp for the `commit_file`. If `read_ict` is false ,this returns the file's
+    /// modification timestamp. If `read_ict` is true, this reads the file's In-commit timestamp.
+    #[allow(unused)]
+    fn commit_file_to_timestamp(
+        &self,
+        engine: &dyn Engine,
+        commit_file: &ParsedLogPath,
+        read_ict: bool,
+    ) -> Result<Timestamp, LogHistoryError> {
+        let commit_timestamp = if read_ict {
+            Self::read_in_commit_timestamp(engine, commit_file)?
+        } else {
+            // By default, the timestamp of a commit is its modification time
+            commit_file.location.last_modified
+        };
+
+        Ok(commit_timestamp)
+    }
+
+    /// Reads the in-commit timestamp for the given `commit_file`.
+    ///
+    /// This returns a [`LogHistoryError::FailedToReadTimestampForCommit`] if this encounters an
+    /// error while reading the file or visiting the rows.
+    ///
+    /// This returns a [`LogHistoryError::InCommitTimestampNotFoundError`] if the in-commit timestamp
+    /// is not present in the commit file, or if the CommitInfo is not the first action in the
+    /// commit.
+    #[allow(unused)]
+    fn read_in_commit_timestamp(
+        engine: &dyn Engine,
+        commit_file: &ParsedLogPath,
+    ) -> Result<Timestamp, LogHistoryError> {
+        debug_assert!(commit_file.is_commit(), "File should be a commit");
+        let wrap_err = |error: DeltaError| LogHistoryError::FailedToReadTimestampForCommit {
+            version: commit_file.version,
+            error: Box::new(error),
+        };
+
+        // Get an iterator over the actions in the commit file
+        let mut action_iter = engine
+            .json_handler()
+            .read_json_files(
+                &[commit_file.location.clone()],
+                InCommitTimestampVisitor::schema(),
+                None,
+            )
+            .map_err(wrap_err)?;
+
+        // Take the first non-empty engine data batch
+        let batch = action_iter
+            .map(|res| res.map_err(wrap_err))
+            .next()
+            .transpose()?;
+
+        // Visit the rows and get the in-commit timestamp if present
+        let in_commit_timestamp_opt = batch
+            .map(|batch| -> Result<Option<i64>, LogHistoryError> {
+                let mut visitor = InCommitTimestampVisitor::default();
+                visitor.visit_rows_of(batch.as_ref()).map_err(wrap_err)?;
+                Ok(visitor.in_commit_timestamp)
+            })
+            .transpose()?
+            .flatten();
+
+        in_commit_timestamp_opt.ok_or_else(|| LogHistoryError::InCommitTimestampNotFoundError {
+            version: commit_file.version,
+        })
```

Comment:
I think we can simplify this?

```suggestion
        let not_found = || LogHistoryError::InCommitTimestampNotFoundError {
            version: commit_file.version,
        };

        // Take the first non-empty engine data batch
        let batch = match action_iter.next()) {
            Some(Ok(batch)) => batch,
            Some(Err(err)) => return Err(wrap_err(err)),
            None => return Err(not_found()),
        };

        // Visit the rows and get the in-commit timestamp if present
        let mut visitor = InCommitTimestampVisitor::default();
        visitor.visit_rows_of(batch.as_ref()).map_err(wrap_err)?;
        visitor.in_commit_timestamp.ok_or_else(not_found)
```